src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Config\Definition;
+
+use Symfony\Component\Config\Definition\Builder\ExprBuilder;
+
+/**
+ * @experimental
+ */
+final readonly class JsonSchemaGenerator
+{
+    public function __construct(private string $outputPath)
+    {
+    }
+
+    public function build(NodeInterface $node, array $schema = []): void
+    {
+        $schema = array_replace_recursive([
+            '$schema' => 'http://json-schema.org/draft-06/schema#',
+            'definitions' => [
+                'param' => [
+                    '$comment' => 'Container parameter',
+                    'type' => 'string',
+                    'pattern' => '^%[^%]+%$',
+                ],
+                'root' => $this->buildSingleNode($node, allowParam: false),
+            ],
+            '$ref' => '#/definitions/root',
+        ], $schema);
+
+        file_put_contents($this->outputPath, json_encode($schema, \JSON_PRETTY_PRINT | \JSON_UNESCAPED_SLASHES | \JSON_THROW_ON_ERROR));
+    }
+
+    private function buildArrayNode(ArrayNode $node): array
+    {
+        $schema = [
+            'type' => ['object', 'array'],
+            'additionalProperties' => $node->shouldIgnoreExtraKeys(),
+            'maxItems' => 0,
```

Comment:
This `maxItems` only accepts an empty array? This seems strange. It's for [framework.validation.auto_mapping.<class-name>](https://github.com/symfony/symfony/blob/d4566b22d633c74b4567a3484ec40b6177d78b1b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L1069), there is no restriction in the array size.

Reply:
The rationale here was that a non-prototyped array means an object – however we'd still allow empty array and treat it the same as empty object as there is no difference in php.

But you're right, it seems validator has the same issue as Doctrine, there is a bunch of [pre-normalization logic](https://github.com/symfony/symfony/blob/d4566b22d633c74b4567a3484ec40b6177d78b1b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L1079-L1102) which makes it impossible to determine the allowed schema.